tla/counterfeit-cycles/counterfeit_cycles.tla

```diff
@@ -0,0 +1,293 @@
+  
+TypesOK ==
+  \* Type-correctness invariant.
+  /\ LedgerTypesOK
+  /\ SubnetsTypeOK
+
+-----------------------------------------------------------------------------
+(***************************************************************************)
+(* Subnet actions                                                          *)
+(***************************************************************************)
+
+SubnetSendTransfer ==
+  (* Honestly sends a transfers cycles from one subnet to another. *)
+  
+  \* Choose two subnets.
+  \E sender, receiver \in SUBNETS: sender /= receiver
+  /\ subnets[sender].balance > 0
+  
+  \* Choose an amount that's within the subnet's balance.
+  /\ \E amount \in 1..subnets[sender].balance:
+  
+  \* Send a transfer to the receiver.
+  /\ subnetMsgs' = Append(subnetMsgs, [
+        type |-> "transfer",
+        from |-> sender,
+        to |-> receiver,
+        amount |-> amount
+     ])
+     
+  \* Subtract amount from sender's balance.
+  /\ subnets' = [subnets EXCEPT ![sender].balance = @ - amount]
+  
+  /\ UNCHANGED<<ledger>>
+
+SubnetDishonestSendTransfer ==
+  (* Maliciously transfers more cycles than the subnet's own balance. *)
+  
+  \E sender, receiver \in SUBNETS: sender /= receiver
+
+  \* To keep the state space bounded, a subnet can make a dishonest
+  \* transfer at most once.
+  /\ subnets[sender].honest
+  
+  \* Send a transfer to the receiver.
+  /\ subnetMsgs' = Append(subnetMsgs, [
+        type |-> "transfer",
+        from |-> sender,
+        to |-> receiver,
+        \* Choose an amount that's greater than the balance.
+        amount |-> subnets[sender].balance + 10
+  ])
+  
+  \* Mark subnet as dishonest to avoid repeating this transfer again.
+  /\ subnets' = [subnets EXCEPT ![sender].honest = FALSE]
+  
+  /\ UNCHANGED<<ledger>>
+
+
+SubnetReceiveTransfer ==
+  /\ Len(subnetMsgs) > 0
+  /\ LET msg == Head(subnetMsgs) IN
+      /\ msg.type = "transfer"
+      /\ subnets' = [
+                         \* Add amount in transfer to the unfinalized balance.
+          subnets EXCEPT ![msg.to].unfinalized[msg.from] = @ + msg.amount,
+                         \* Add message to ledger outgoing queue.
+                         ![msg.to].msgsToLedger = Append(@, msg)
+         ]
+      \* Remove message from subnet messages.
+      /\ subnetMsgs' = Tail(subnetMsgs)
+      /\ UNCHANGED<<ledger>>
+      
+LedgerPoll ==
+  \E s \in SUBNETS: Len(subnets[s].msgsToLedger) > 0
+    /\ LET msg == Head(subnets[s].msgsToLedger) IN
+      \* Send message to ledger.
+      /\ ledger' = [ledger EXCEPT !["msgs"] = Append(@,
+          [type |-> "transfer", from |-> msg.from, to |-> msg.to, amount |-> msg.amount]
+        )]
+      \* Remove message from the queue.
+      /\ subnets' = [
+            subnets EXCEPT ![s].msgsToLedger = Tail(@)
+         ]
+      /\ UNCHANGED<<subnetMsgs>>
+
+SubnetReceiveApprove == 
+  /\ Len(subnetMsgs) > 0
+  /\ LET msg == Head(subnetMsgs) IN
```

Comment:
Could you be omitting some valid behaviors here because you impose an ordering? In particular, you only ever process the head of the `subnetMsgs` queue, but there's no guaranteed order in which different subnets would process these responses, and there's also no delivery order between messages sent by the other subnets and the ledger responses (actually, the interface spec doesn't even guarantee the delivery order for responses even between the ledger and the same subnet, though our implementation generally does). The same is actually true for the cycles-bearing messages between subnets (the receive order between subnets is not defined).

Unless you can justify why a queue is enough (which would actually likely be beneficial from an analysis perspective, as it would greatly decrease the number of possible actions), I'd suggest switching to sets. There are certain standard techniques that seem related to this (the queue could potentially be a so-called "prophecy variable" of how you are going to perform actions), but it can be a bit finicky.

Reply:
> The problem when I used a set was starvation - TLA would identify a sequence where a certain message is added to subnetMsgs but is never processed. Any ideas on how to address this?

I'd have to look at the concrete trace, but it sounds suspicious to me that you had a starvation problem (I guess TLC complained of deadlock?) with sets, but not with lists. In general a set should be more "lenient", in the sense that you allow more actions and are thus less likely to deadlock.

If you can recreate the problem by running TLC from the command line, you should get a `counterfeit_cycles_TTrace_<something>.tla` file somewhere (probably also if you use the toolbox, but I'm not really familiar with how that handles problems) that contains the counterexample and that I can have a look at.

> I thought of trying to model the IC more precisely by having a separate queue per sender/receiver, but was worried of the complexity of that approach.

Using a set should overapproximate the real behaviors (I guess not a surprise). This is sound for safety properties (all of your properties except for the ledger detecting a malicious subnet are safety). The only downside is that the analysis might perform a lot worse, but premature optimization and all that, so I'd start with a set.

There's already [a model of streams](https://github.com/dfinity-lab/formal-models/blob/b5a7e6390f7140ffa827a55410cb681c2e2b2e73/tla/canister-migration/Migration.tla#L271) per subnet pair. It's far too detailed for your purposes, but I thought I'd mention it if useful for inspiration (I'd still try sets first, though)

Reply:
One of the problems with using a set for `subnetMsgs` is that transfers messages are not unique and can therefore be lost if the same transfer (e.g. of 1 cycle from "s1" to "s2") are done in a row.

I've made the following changes to address this:

1. I added an `id` field to every message. Similar transfers will have different IDs and therefore both remain represented in the set of `subnetMsgs`.
2. I added a `MAX_TRANSFERS` constant, which limits the number of transfers that can be done. Otherwise, the introduction of the `id` field would make the state space infinite AFAIU. With that introduction, I had to turn off the check for dead locks as TLC assumes we've hit a deadlock once `MAX_TRANSFERS` has been made.
3. I changed the `msgs` that the ledger receives from a queue to a set of queues, where there's a queue per subnet. This is more representative of what happens on the IC. This isn't strictly required but I thought it's a more accurate representation.

Reply:
I went ahead with the idle transition and re-enabled the deadlock check. Thanks for the tip!